lms/djangoapps/courseware/user_state_client.py

```diff
@@ -222,8 +226,19 @@ def set_many(self, username, block_keys_to_state, scope=Scope.user_state):
                 current_state.update(state)
                 num_fields_after = len(current_state)
                 student_module.state = json.dumps(current_state)
-                # We just read this object, so we know that we can do an update
-                student_module.save(force_update=True)
+                try:
+                    with transaction.atomic():
+                        # Updating the object - force_update guarantees no INSERT will occur.
+                        student_module.save(force_update=True)
+                except IntegrityError:
+                    # The UPDATE above failed. Log information - but ignore the error.
+                    # See https://openedx.atlassian.net/browse/TNL-5365
+                    log.warning("set_many: IntegrityError for student {} - course_id {} - usage key {}".format(
```

Comment:
We have transactions that can potentially update dozens of state fields -- will this translate into a separate transaction for each one?